[체스 - 3, 4 단계] 콜리(김건우) 미션 제출합니다 🥦

coli-geonwoo opened the pull request.

coli-geonwoo committed:
docs: 완료한 기능 구현 사안 체크

coli-geonwoo committed:
docs: 3단계 기능 구현 사안 작성

coli-geonwoo committed:
refactor: 진영 전환 책임을 Team 객체와의 협력으로 리팩터링

coli-geonwoo committed:
feat: 각 Piece에 점수 인스턴스 변수 추가

coli-geonwoo committed:
refactor: 점수를 VO로 래핑

coli-geonwoo committed:
refactor: RowPosition, ColumnPosition POOL 생성

coli-geonwoo committed:
refactor: getScore 반환 타입 변경

double > Score

coli-geonwoo committed:
feat: Score 연산 메서드 및 eq&hc 추가

coli-geonwoo committed:
feat: 폰인지 확인하는 메서드 재정의

coli-geonwoo committed:
feat: 점수를 계산하는 ScoreCalculator 구현

coli-geonwoo committed:
feat: chessBoard가 인스턴스 변수 calculator를 통해 점수를 반환하도록 조합

coli-geonwoo committed:
feat: chessBoard에서 해당 팀의 킹이 살아있는지 알 수 있다

coli-geonwoo committed:
feat: 체스게임 인터페이스화

상태에 따라 RunningGame과 EndGame으로 이분화

coli-geonwoo committed:
rename: 제3의 팀 이름 변경

EMPTY > NONE

coli-geonwoo committed:
feat: chessGame에서 승리 팀 확인기능 구현

coli-geonwoo committed:
feat: 이동 시 game 최신화 기능 추가

coli-geonwoo committed:
fix: 잘못된 우승팀 반환 오류 개선

coli-geonwoo committed:
test: 무승부/승리 상황 시 우승팀 반환 기능 구현

coli-geonwoo committed:
docs: 기능구현 사안 체크

coli-geonwoo committed:
feat: 점수 및 경기결과 출력 outputView 메서드 구현

coli-geonwoo committed:
feat: StatusCommand 구현

coli-geonwoo committed:
feat: ChessGame의 isEnd 구현

coli-geonwoo committed:
feat: execute 반환형을 ChessGame으로 수정

coli-geonwoo committed:
refactor: game의 상태패턴을 제거

game 단일 클래스로 로직 변경

coli-geonwoo committed:
docs: 기능구현사안 체크

coli-geonwoo committed:
chore: db 환경설정

coli-geonwoo committed:
refator: toKey 접근제어자 변경

private static > public static

coli-geonwoo committed:
feat: DB저장을 위한 getter 추가

coli-geonwoo committed:
refactor: 폰 생성로직 변경

Pawn.blackPawn() > Pawn.of(Team.BLACK)
Pawn.whitePawn() > Pawn.of(Team.WHITE)

coli-geonwoo committed:
feat: DB와 도메인간 pieceMapping을 도와주는 PieceMapper 구현

coli-geonwoo committed:
feat: DB와 도메인간 teamMapping을 도와주는 PieceMapper 구현

coli-geonwoo committed:
feat: DB연결 로직 구현

coli-geonwoo committed:
feat: GameDao naive하게 구현

coli-geonwoo committed:
feat: 컨트롤러에 DB연결

coli-geonwoo committed:
style: 컨벤션에 맞게 코드 수정

coli-geonwoo committed:
refactor: dao 역할 분리

boardDao > 보드 저장 및 불러오기
turnDao > 턴 저장 및 불러오기

coli-geonwoo committed:
feat: DB스키마 추가

coli-geonwoo committed:
style: 코드 컨벤션에 맞게 수정

coli-geonwoo committed:
remove: db테스트용 객체 삭제

coli-geonwoo commented:
안녕하세요 찰리 🙌  콜리입니다. 🥦 

이번 3-4단계는 DB를 처음 다루어보는 제겐 많이 힘든 미션이었습니다.  DB저장과정까지 모두 구현하였으나 아직 클린코드화 + 확장성을 고려한 DB설계까지는 신경쓰지 못했습니다. 그러나, 이 리팩터링까지 완료한 이후에 찰리에게 리뷰를 요청하면 시간이 너무 지체될 것 같아 우선 리뷰 요청을 드립니다.  

**DB 스키마**는 다음과 같습니다.

```sql
create table piece_type
(
    type varchar(20) not null
        primary key
);

create table team_color
(
    color varchar(20) not null
        primary key
);

create table board
(
    position       varchar(20) not null
        primary key,
    distinct_piece binary(1) not null,
    piece_type     varchar(20) null,
    team           varchar(20) null,
    constraint piece_type
        foreign key (piece_type) references piece_type (type),
    constraint team
        foreign key (team) references team_color (color)
);

create table turn
(
    team varchar(20) null,
    constraint turn
        foreign key (team) references team_color (color)
);

```

몇가지 고민사안은 다음과 같습니다.

**> 1.  Dao 코드의 경우 TDD를 어떻게 적용해야할지 고민이 생겼습니다.**


제가 이해한 테스트 중심 개발원칙의 핵심은 
`테스트를 검증방식이 아닌 하나의 설계방식으로 활용`하는 것입니다. 
<br>
그러나 dao의 경우 그 역할이 `보드를 불러온다. 보드를 저장한다` 처럼 이미 어느정도 정해져 있고
 그 기능을 검증하는 것(predfined method 기능을 검증하는 것)에 지나지 않는다는 생각이 들었습니다.
또한, 실제 DB를 사용한다는 점도 마음에 걸렸습니다.  
<br>
=> 이에 테스트 더블이란 개념을 검색을 통해 알게되었고 그중 stub을 다음 리팩터링에 적용해볼 생각입니다. 
<br>

DB와 관련하여 **찰리에게 여쭙고 싶은 사안**은 다음과 같습니다.

> Dao 또한 테스트를 기반으로 한 `검증`이 아닌 `설계`가 가능한가요?

<br>

> DB 설계의 경우 DBA의 역할이라 생각되는데 서버 개발자로서 DB에 대해 책임을 져야하는 범위가 현업에서는 어느정도인지 찰리의 경험들을 여쭙고 싶습니다 (DB 관리에 익숙치 않아 제가 당황한 부분도 있어 질문드립니다. 😓 )

<br>

> 도메인에서 db로의 전달을 위해 getter를 이곳저곳 넣어주는 상황이 생겼는데, 다른 방안을 생각해도 저는 생각이 나지 않았습니다. 혹시 이를 우회하면서 도메인의 정보를 dao가 전달할 수 있는 방안이 있을까요? 🤔 


---

## 이외에 드리고 싶은 설계할 때의 고려했던 제 생각들

#### ChessBoard에서 ScoreCalculator를 가지고 있는 이유 

-  정보 전문가 패턴에 따라 Score를 들고 있는 Piece에 대한 정보를 가장 많이 알고 있는 Board에서 점수를 반환하는 것이 어색하지 않다고 생각했습니다.
---
#### ScoreCalculator에서 ColumnPosition getter를 활용한 이유

-  각 열에 Pawn이 몇개인지 확인하고 2개 이상이라면 점수를 깎는 메서드가 필요했습니다.
- getter 사용하지 않고도 구현할 수 있었으나 코드가 생각보다 많이 길어졌습니다.
-  columnPosition 별로 Stream의 grouping 기능을 활용하면 이 기능을 훨씬 빠르고 명확하게 구현가능했습니다.
- Dao를 위해서라도 ColumnPosition getter가 필요했고 도메인에서 이를 활용하는 것은 별개의 영역이지만 리소스 트레이드 오프를 생각했을 때, 더 짧고 명확한 코드의 이점이 크다 생각했습니다

```java
 private Score calculateMinusPawnScore(List<Position> pawnPositions) {
        Score result = ZERO_SCORE;

        Map<ColumnPosition, Long> group = pawnPositions.stream()
                .collect(groupingBy(Position::getColumnPosition, counting()));

        for (ColumnPosition column : ColumnPosition.POOL.values()) {
            result = result.add(oneColumnPawnScore(group.getOrDefault(column, 0L)));
        }
        return result;
    }
```

---

#### King이 점수(0점)를 들고 있는 것에 대하여

King과 NullPiece는 점수를 가지고 있지 않아야 합니다. King이 잡히면 게임이 끝나고 NullPiece는 널오브젝트이기 때문입니다. 이를 위해 King과 NullPiece에서 점수를 호출하면 오류를 낼 생각으나 생각을 바꾸어 NullPiece에서만 오류코드를 던지기로 했습니다.

그 이유는 다음과 같습니다.
코드의 외부 사용자가 관심있어 하는 것 혹은 도메인이 목적으로 하는 기능은 `게임의 팀별 점수를 구하는 것`이라 생각했습니다.
사용자 입장에서 King 점수 계산의 오류처리를 하는 것은 관심영역 밖의 일에 대한 부담을 호출 코드에게 이전하는 일이라 생각했습니다. 

---

리팩터링 양이 조금 많아 찰리가 리뷰과정에서 코드 트래킹이 어려우실 수도 있을 것 같아 죄송한 마음입니다. 리뷰 남겨주시면 저도 빠른 시일 내에 더 완성된 코드로 리뷰 요청 드리겠습니다!

매번 좋은 리뷰에 감사드립니다 🙇

Gomding commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -1,7 +1,9 @@
 package chess.view;
```
게임 시작 시 `status` 명령어도 안내가 되면 좋겠어요 :)
현재는 사용자가 `status` 명령어를 인지하지 못하겠네요! 😃 

![image](https://github.com/woowacourse/java-chess/assets/57378410/59c60c0a-05e5-4673-92b7-51b53a9c0625)

coli-geonwoo replied:
사용자 경험에 대한 생각을 하지 못했네요 😅  반영해주었습니다!

Gomding commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+    volumes:
+      - ./db/mysql/data:/var/lib/mysql
+      - ./db/mysql/config:/etc/mysql/conf.d
+      - ./db/mysql/init:/docker-entrypoint-initdb.d
```
(절대 필수가 아닌 반영..이지만... 해주면 앞으로 미션에서 리뷰어들이 좋아하는 부분..)
docker 디렉터리를 보면 `init` 이 있을 거에요!
아래와 같이 sql 파일을 생성해주면
`docker-compose` 명령어로 컨테이너 생성 시 테이블을 함께 생성할 수 있습니다 😃 

![image](https://github.com/woowacourse/java-chess/assets/57378410/bfb914fd-8763-4d9d-a4a7-dfc1bf5f8552)

coli-geonwoo replied:
리뷰어의 편의는 곧 리뷰의 질과도 연결된다고 생각해요!! 친절하게 사진까지 첨부해서 알려주셔서 감사합니다 🙇 찰리의 꿀팁 덕분에 앞으로 db를 다루는 미션에서 제가 원하는 방향의 리뷰를 받을 수 있을 것 같아요 🙌

Gomding commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -10,6 +11,7 @@
 import static chess.view.CommandMapper.START;
 
 public class ChessGameController {
+    private static final GameDao gameDao = new GameDao();
```
ChessGameController가 GameDao 를 알고있군요! 🤔 
Controller 가 너무 구체적인 사항을 알게된 것이 아닐까요..!
`게임이 언제 저장되야 하는가?` 라는 부분이 Controller 에 노출된 것으로 보입니다 :)
더 적절한 객체에게 해당 부분을 맡겨보죠 :)

Gomding commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -0,0 +1,62 @@
+    private static final String DATABASE = "chess";
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "root";
```
`BoardDao`, `GameDao` 각각에 DB 접속 정보가 중복 되어있네요 :)
해당 부분을 관리하는 객체가 있다면 좋을것같아요 😃

Gomding commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -0,0 +1,62 @@
+
+    public Connection getConnection() {
+        try {
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
```
Connection 을 가져오는 부분도 `BoardDao`, `GameDao` 각각 동일하게 존재하고 있어요!
해당 부분도 적절한 객체가 있으면 중복을 깔끔하게 제거할 수 있겠군요 :)

만약 DB 접속 정보나 Connection 가져오는 로직이 변경되면 
현재는 각각의 Dao 를 모두 수정해줘야하고
이 과정에서 문자열 오타나 코드 실수가 나올 수 있으니까요!
적절한 객체가 있다면 바뀌더라도 최소한의 수정으로 처리할 수 있습니다 😃 

접속 정보나 커넥션 가져오는 부분을 캡슐화하는 목적에서요!

coli-geonwoo replied:
커넥션과 관한 중복된 책임을 ConnectionGenerator로 넘겨주었습니다!

Gomding commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -0,0 +1,62 @@
+
+    private boolean canFindRunningGame() {
+        final var query = "SELECT * FROM board WHERE distinct_piece = 1;";
+        try (final var connection = getConnection();
```
try-with-resources 사용 👍 

커넥션을 자연스럽게 자원 해제해주겠군요

Gomding commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -0,0 +1,62 @@
+        turnDao.saveTurn(game);
+    }
+
+    public ChessGame findGame() {
```
1. 메서드 이름은 조회에 가까운데 내부에는 생성하는 로직이 있어요 🤔 
2. 기존 진행하던 게임이 있으면 ~ 불러오기
    기존 게임이 없으면 ~ 새로 생성
    DAO의 풀네임을 보면 `데이터 접근 객체 (Data access object)`로 데이터 접근에 대한 책임을 가지는것을 추측할 수 있어요 :)
    그럼 위 책임을 수정하고 싶을 때 개발자는 도메인 VS DAO 어느쪽을 먼저 보게될까요?
    DAO 에게는 좀 더 심플한 행위만 있어도 된다고 생각해요 😃

coli-geonwoo replied:
>그럼 위 책임을 수정하고 싶을 때 개발자는 도메인 VS DAO 어느쪽을 먼저 보게될까요?

찰리가 가정한 상황이 제 코드에 어떤 문제가 있는지 명확히 알게 해주었어요.
> 배운점1 : DAO는 데이터 접근에 대한 책임만을 가진다
> 배운점2: 코더가 가정한 상황은 사용자에게 예측가능하지 못한 결과를 넘겨줄 수 있다

**DAO에는 조회의 기능만을 남겨주었고, 새로운 게임으로 대체하는 로직은 DBService가 수행하도록 해주었습니다.**

Gomding commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -0,0 +1,62 @@
+            final var resultSet = preparedStatement.executeQuery();
+            return resultSet.next();
+        } catch (final SQLException e) {
+            throw new RuntimeException("현재 실행중인 게임 조회과정에 오류가 생겼습니다.");
```
RuntimeException으로 변경하여 던져주셨군요 😃 
어떤 이유에서 예외를 변경하여 다시 던져주셨을까요?

coli-geonwoo replied:
사실 오류 선정에 있어서는 명확한 근거가 없습니다. 그러나 RuntimeException으로 처리를 하게 된 과정을 설명드리면 다음과 같습니다.

- SQLException이 checkedError라 필수적인 예외처리가 필요했습니다.
- SqlException 오류처리에 있어 중대한 에러사안이므로 개발자에게 알려야 겠다는 생각에 새로운 오류를 던져주고자 했습니다.
- DB 처리는 런타임에 이루어진다는 점에서 RuntimeException에 메세지를 담아 던져주었습니다.

Gomding replied:
UncheckedException으로 변환하여 던져준 것은 💯 점이라 생각해요!
제가 생각하는 근거를 공유드리자면..!

> SqlException 오류처리에 있어 중대한 에러사안이므로 개발자에게 알려야 겠다는 생각에 새로운 오류를 던져주고자 했습니다.

만약 중대한 에러 사항이라서 개발자에게 알려야한다면 
`RuntimeException` 과 같은 `UncheckedException` 보다는 `SQLException` 과 같은 `CheckedException`이 더 어울릴 거에요!
CheckedException은 try-catch 에서 처리하거나 throws 를 하도록 강제하니까요!

`SQLException`을 `UncheckedException`로 변환하여 던져야할까? 를 고민해본다면 다음 질문들을 고민해볼 수 있어요!
- 호출한 쪽에서 발생한 에러에 대해 복구가능한가? (try-catch 로 잡아서 정상적인 상황으로 복구가 가능한지?)
    - 예를들면) 쿼리를 잘못 작성하여 발생한 에러인데 바깥에서 복구가 가능한가?
    - DB 접속 정보를 잘못 작성하여 접속이 불가능한데 바깥에서 catch 하여 복구가 가능한가?
- 바깥에서 SQLException을 알아야할까?
- 의미없는 throws 가 계속되지 않을지?

coli-geonwoo replied:
찰리의 리뷰를 듣고, 토비의 스프링 4장에 있는 예외 파트를 한번 읽어보았어요. 관련 레퍼런스가 많아 찰리가 어떤 의도로 다음 질문들을 주신 것인지 참뜻을 이해할 수 있었습니다.

>호출한 쪽에서 발생한 에러에 대해 복구가능한가? (try-catch 로 잡아서 정상적인 상황으로 복구가 가능한지?)   

SQLException 같은 경우 애플리케이션 레벨에서는 복구가 불가한 케이스가 대부분이기 때문에 신경쓰지 않도록 해주는 것이 바람직합니다.

---

> 바깥에서 SQLException을 알아야 할까? / 의미없는 throws 가 계속되지 않을지?      

SQLException을 그대로 던지는 예외처리 회피를 하게 된다면 dao를 다루는 서비스 계층의 메서드마다 thorws SQLException이 붙어야 할 것이고, 이는 복구가 불가한 외부 메서드들에 의미없는 thorws가 계속됨을 의미합니다. 

그렇기에 checkedException인 SQLException을 RuntimeException으로 포장하여 던져주는 것이 여러 측면에서 바람직합니다 😄

Gomding replied:
헉 이미 찾아서 읽었으셨군요 💯 💯 💯 💯 
훌륭합니다~!!

Gomding commented on src/main/java/chess/dao/DBSchema.sql:
```diff
@@ -0,0 +1,31 @@
+create table piece_type
```
piece_type, turn 의 경우 상수로 보여요
꼭 DB에 있어야하는 정보일지 고민해보아요! 🚀

coli-geonwoo replied:
piece_type의 경우 외래키 설정을 위해 만들어준 table이고 turn의 경우 현재 board table에 turn 정보가 없어 따로 분리된 turn table입니다.

piece_type은 DB에 없어도 되는 테이블로 생각됩니다.
그러나, turn은 다시 game을 db에서 불러올 때 필요한 테이블이라 생각합니다!

Gomding replied:
앗 코멘트를 잘못 썻네요;;

말씀드리려던건 piece_type 과 team_color 였습니다..!!
외래키로 DB에 값을 강제할 수 있다~! 라는 장점이 있긴하지만

그건 도메인 레벨에서 충분히 처리할 수 있다고 생각해요 😄

Gomding commented on src/main/java/chess/dao/DBSchema.sql:
```diff
@@ -0,0 +1,31 @@
+    constraint piece_type
+        foreign key (piece_type) references piece_type (type),
+    constraint team
+        foreign key (team) references team_color (color)
```
foreign key는 어떤 이유로 사용해주셨을까요!

coli-geonwoo replied:
board에서 들어온 데이터가 사전의 정의된 piece_type과 team에서 벗어나는 경우를 방지하고자 했습니다 💭

Gomding replied:
좋은 의도라고 생각해요!
하지만 해당 부분은 도메인에서 충분히 검증할 수 있는 부분이 아닐까요? 🤔 

외래키의 경우 잘 사용하면 좋지만
대부분의 경우 독으로 작용하는 경우를 많이 봤어요 😭 

DB를 마이그레이션하거나 버전업 등의 작업을 할 때
외래키로 인해 고통받은적이 많습니다..

Gomding commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -20,13 +22,16 @@ public ChessGameController(InputView inputView, OutputView outputView) {
-        } while (!command.isEnd());
+        } while (!command.isEnd() && !game.isEndGame());
+
+        gameDao.saveGame(game);
```
> 애플리케이션을 재시작하더라도 이전에 하던 체스 게임을 다시 시작할 수 있어야 한다.

해당 부분을 위해 여기서 저장하신 것으로 보여요..!
만약 게임에서 말을 옮길때마다 체스 데이터를 옮겨준다면 해당 부분이 필요하지 않을것 같아요 😃 
게임 중 컴퓨터 오류나 천재지변으로 인해 게임이 강제 종료되더라도 데이터가 남아있을것이구요 :)

무엇보다도 체스말을 움직인다는건 상태가 변한다는 것이고
이것을 저장해주는것이 더 자연스럽다고 생각합니다 :)
(무조건 그런것은 아니지만요 ..! 기획에 따라 달라질 수는 있다고 봅니다 ㅎㅎ)

coli-geonwoo replied:
move마다 상태를 저장할 수 있도록 리팩터링 해주었습니다!

Gomding commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -20,13 +22,16 @@ public ChessGameController(InputView inputView, OutputView outputView) {
-        } while (!command.isEnd());
+        } while (!command.isEnd() && !game.isEndGame());
+
+        gameDao.saveGame(game);
```
현재 체스게임 정보가 저장되고 있지않아요 😢 
원인을 파악하고 해결해주세요! 🙏 
(Insert 되고 있는곳이 현재 존재하고 있을까요? 😱 )

coli-geonwoo replied:
local DB에는 console로 insert문을 넣어주었는데, 이것이 현재 코드파일에 적혀져 있지 않네요 😭 
로컬에서만 잘된다는 걸 확인하고 찰리에게 리뷰 요청 보낸 점 죄송합니다 🙇‍♂️ 

docker/mysql/init폴더에 다음 두 sql파일을 첨부했습니다!!
create_table.sql : 초기 테이블 생성 sql
insert_data.sql : 초기 데이터 삽입 sql

Gomding replied:
헉 새로운 보드를 저장하는 부분도 체스게임 코드내부에 있어야한다고 생각해요 😱 

만약 온라인 체스게임으로 전환한다면
현재 구조에서 가능할까요?

DB를 위한 sql 파일은 단순히 서비스의 환경설정을 위한 것이고
(또는 테스트를 위해 활용)
모든 DB에 데이터 저장/수정/삭제/조회는 코드 내에서 처리해주세요!

coli-geonwoo replied:
아하...DDL을 제외한 초기 데이터 설정도, 코드 내부에 있어야 하는군요 😢  boardDao와 turnDao에 각각 initialize를 통해 첫번째 call에서만 db를 초기화할 수 있도록 설정해주었습니다!! 다만 떠오르는 구현이 많지 않아 sql로 db 행이 존재하는지 조회하였는데 db접근에도 시간이 걸리는 만큼 이것이 좋은 구현인지는 모르겠습니다...

```java
private boolean isFirstCall() {
        final var query = "SELECT COUNT(*) AS 'CNT' FROM board";
        try (final Connection connection = connectionPool.getConnection();
             final PreparedStatement preparedStatement = connection.prepareStatement(query);
             final ResultSet resultSet = preparedStatement.executeQuery()) {
            connectionPool.returnConnection(connection);
            return !resultSet.next();
        } catch (final SQLException e) {
            e.printStackTrace();
            throw new RuntimeException("조회 기능 오류");
        }
    }

    private void initializeBoard() {
        Set<String> positions = Position.POOL.keySet();
        System.out.println(positions);
        for (String position : positions) {
            updateOnePosition(position);
        }
    }
```

Gomding replied:
이건 해당 부분 보면서 코멘트 남겨볼게요!

Gomding commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -0,0 +1,43 @@
+package chess.dao;
```
> 제가 이해한 테스트 중심 개발원칙의 핵심은
> 테스트를 검증방식이 아닌 하나의 설계방식으로 활용하는 것입니다.
> 
> Dao 또한 테스트를 기반으로 한 검증이 아닌 설계가 가능한가요?

이 질문에서 TDD를 하나의 설계 방식으로 활용한다는 것이 어떤것인이 감이오지 않았아요..! 😢 
어떤 방식인지 예시를 조금만 들어주시면 좋겠어요..!! 🙏

coli-geonwoo replied:
제가 의도했던 것은 TDD 사이클을 지키며, 객체간의 책임과 역할(interface)을 테스트를 통해 인식하는 과정입니다.

제가 생각하는 TDD를 통한 설계 방식
- 1단계  - 통과하지 않는 테스트 작성 : 구현 이전에 객체의 책임을 인식하기 위해 assert파트만 작성
- 2단계 - 메서드 기능 구현
    -  메서드를 구현하다가 새로운 역할이 필요하면 새로운 객체에 대한 테스트 작성
- 3단계 - 테스트 작성 마무리 및 통과하는지 검증

이렇듯 테스트 작성 과정에서 "무엇을 할 것인가?"에 대한 책임을 명확히 인식하고 이후에 구현으로 돌아와 "어떻게 할 것인가?"를 생각하는 과정을 말합니다.

그런데 DAO의 경우, TDD를 적용하기 힘들었는데 그 이유는 다음과 같습니다.

- 이미 책임이 나누어진 느낌이었습니다. boardDao나 다른 객체들의 책임이 '보드를 가져온다' / '보드를 저장한다.'처럼 테스트 작성 전에 이미 구분이 되어 있는 느낌이었습니다.
- 실제 DB를 적용한다는 점이 마음에 걸렸습니다. 테스트 로직 때문에 DB를 건드린다는 것이 좋지 않아보였습니다.

이런 점들을 보완하기 위해 이번에는 FakeBoardDao를 테스트용 객체로 만들어 한번 `테스트 더블`의 개념을 적용해보았는데요. 많이 시간을 쏟았지만 이것이 많는 테스트 더블 방식인건지 아직 애매하게 와닿는 것 같습니다 😢

Gomding replied:
`DAO 의 경우 책임이 명확하게 구분되어 있다고 느꼈다` 라고 이해했는데 맞을까요!

DAO 의 경우 데이터 접근을 위한 객체로
그 책임을 더 이상 나누기는 힘들거에요 ㅎㅎ! (예외의 경우도 있겠지만요)
콜리가 느낀것이 맞다고 생각합니다~!
그리고 더 많은 책임을 주는것도 어색하구요

> 실제 DB를 적용한다는 점이 마음에 걸렸습니다. 테스트 로직 때문에 DB를 건드린다는 것이 좋지 않아보였습니다.

이 부분도 FakeBoardDao 를 잘 만들어주셨다고 생각해요 
해당 부분에 고민해보셨으면 하는 것이 있어 코멘트 남겨놓겠습니다!

Gomding commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -0,0 +1,43 @@
+package chess.dao;
```
> DB 설계의 경우 DBA의 역할이라 생각되는데 서버 개발자로서 DB에 대해 책임을 져야하는 범위가 현업에서는 어느정도인지 찰리의 경험들을 여쭙고 싶습니다 (DB 관리에 익숙치 않아 제가 당황한 부분도 있어 질문드립니다. 😓 )

제가 DBA가 있는 회사에 다녀본적이 없어 확답을 드리긴 어렵지만 ㅎ...
주변의 이야기를 많이 들어봤을 때는
DBA가 DB 테이블까지 설계해주지 않습니다 😄 
보통 DBA는 테이블의 구조적 결함, 쿼리 사용 가능 여부(성능 측면에서 체크) 등을 해주는 것으로 알아요
(그 외에도 DB관련하여 많은 일들을 하지만.. 서버개발자와 관계있는 부분만 말씀드리면요 ㅎㅎ)
테이블 설계는 해당 도메인을 담당하고 잘 알고있는 `서버 개발자` 가 하는 것이 보통입니다!

coli-geonwoo replied:
그렇군요!! 간단히 드린 질문인데도 정성스레 답변해주셔서 정말 감사합니다 👍 
db설계와 관련해서도 아직 많이 부족한 만큼 정진해보도록 하겠습니다. 💪

Gomding commented on src/test/java/chess/dao/GameDaoTest.java:
```diff
@@ -0,0 +1,43 @@
+package chess.dao;
```
> 도메인에서 db로의 전달을 위해 getter를 이곳저곳 넣어주는 상황이 생겼는데, 다른 방안을 생각해도 저는 생각이 나지 않았습니다. 혹시 이를 우회하면서 도메인의 정보를 dao가 전달할 수 있는 방안이 있을까요? 🤔

이 부분은 어쩔 수 없는 부분이라고 생각해요 :) 제일 간단하기도 하구요!

getter 는 무조건 피하는것보다 필요할 때 사용하는것이 맞다고 생각합니다!
getter 사용을 피하라는 이유는 도메인 로직상에서 getter를 사용하면
객체의 책임이나 필드가 바깥에 노출되는 상황이 생겨서 그렇다고 생각해요 (캡슐화가 깨짐)

즉 도메인의 값을 DAO에 넘겨주기위한 getter 사용은 위의 사항과 관련이 없기때문에
괜찮다고 생각해요 🙏

Gomding commented on src/main/java/chess/dao/DBSchema.sql:
```diff
@@ -0,0 +1,31 @@
+create table piece_type
```
sql 파일은 따로 뺴주는 것이 좋습니다 :)

`java-chess/src/main/resource` 소스코드 외 프로젝트와 관련된 파일들은 보통 해당 디렉터리에 위치시킵니다 😄 

`java-chess/src/main/resource/적절한디렉터리/DbSchema.sql` 이렇게 해주면 다른 개발자가 눈치채기 좋을것같아요~ (README.md 에도 적어주면 친절함의 화룡점정 입니다)

coli-geonwoo replied:
반영해주었습니다!

Gomding replied:
잘 반영해주셨네요~

README에는 해당 프로젝트에 신규로 참여한 개발자라도 
README만 읽고 실행할 수 있는 정도로 작성해주시면 좋아요~

예를들면

```md
### 실행 방법

`docker-compose -p chess up -d` 실행하여 DB 를 설정한다

ChessApplication 의 main 메서드 실행
```

coli-geonwoo replied:
찰리의 예시를 보고 더 명확히 안내해주어야 함을 느꼈습니다!! 
README.md 파일에 반영해주었습니다!!

```
### 실행 방법

1단계 : `docker-compose -p chess up -d` 를 실행하여 DB를 설정한다
2단계 : Chessapplication의 메서드를 실행한다
- board : 체스 게임의 보드 현황 저장
- turn : 체스 게임의 턴 저장

```

Gomding reviewed (changes requested):
안녕하세요 콜리!
어려운 단계였음에도 고민과 함께 깔끔하게 구현해주셨네요 👍 
고생하셨습니다 🎉 

몇 가지 코멘트 남겼으니 확인부탁드려요! 🙏 
(이번에는 질문에 대한 답이 코멘트 후반부에 위치해있습니다 🙇 )

궁금한 점 있으면 언제든 DM 부탁드려요!

coli-geonwoo committed:
chore : sql파일 디렉토리 변경

chess/dao > resource

coli-geonwoo committed:
chore: 3/4단계 1차 개선사안 TODO작성

coli-geonwoo committed:
chore: 메서드 순서 변경

coli-geonwoo committed:
refactor : 커넥션 담당 객체 분리

coli-geonwoo committed:
refactor : 게임 결과 커멘드 메시지 안내

coli-geonwoo committed:
fix: connect 오류 개선

coli-geonwoo committed:
refactor: service layer로 트랜잭션 캡슐화

coli-geonwoo committed:
style: 메서드 순서 변경

coli-geonwoo committed:
feat: 초기 table 생성 및 데이터 삽입 sql파일 추가

coli-geonwoo committed:
refactor: turnDao에서 DB연결 책임 분리

coli-geonwoo committed:
refactor: try-with-resource 를 통한 자원 해제

coli-geonwoo committed:
rename: GameDao.java > DBService

coli-geonwoo committed:
chore: 커넥션 관리 객체 디렉토리 변경

dao > database

coli-geonwoo committed:
refactor: 새로운 게임으로 리셋 기능을 시작지점으로 이전

coli-geonwoo committed:
test: Fake객체를 활용한 boardDao test 작성

coli-geonwoo committed:
rename: 의도가 잘 드러나도록 메서드명 변경

setNewGame > saveNewGame

coli-geonwoo committed:
docs: 3-4단계 1차 리뷰 개선사안 작성

coli-geonwoo commented:
안녕하세요 찰리 🙌  콜리입니다 🥦 

3-4단계 1차 리뷰 개선사안을 전달드립니다.

---

### 1차 리뷰 개선사안
- sql 파일 첨부
  - docker/mysql/init/create_table.sql : 초기 테이블 생성
  - docker/mysql/init/insert_data.sql : 초기 데이터 생성

- service layer 생성
  - ChessGameService : db와 domain간의 상호작용 캡슐화
  - DBService : boardDao와 TurnDao간의 상호작용 캡슐화

- DB의 생성 로직 제거 : 단순 접근 책임을 가지도록 리팩터링
- connetion 관리 객체 분리
- status 안내 메시지 추가
- FakeBoardDao 객체를 통한 boardDao 테스트

---

이번 리팩터링에서 신경썼던 점은 다음과 같습니다.
- FakeBoardDao를 통해 테스트 더블 개념 이해하기
- Service layer의 DB-Domain간의 interface적 역할 이해하기
- 더 좋은 DB설계 고민해보기

하지만 아직 경험의 부족으로 온전한 이해까지는 스스로 많이 부딪혀보는 시간이 필요하다 사료됩니다. 리뷰 남겨주시면 스스로 부족한 점이 무엇이었나 되묻고 한번 더 생각해보겠습니다 🙇‍♂️

Gomding commented on docker/db/mysql/init/insert_data.sql:
```diff
@@ -0,0 +1,69 @@
+INSERT INTO 'board' (position, distinct_piece, piece_type, team)
```
init 폴더 안에 sql 파일이 어떤 순서로 실행될지 예상할 수 있을까요? 🤔 

+) piece_type 과 team_color 값이 없어서 해당 sql은 실행하고 싶어도 실행이 불가능했어요 😢

coli-geonwoo replied:
>init 폴더 안에 sql 파일이 어떤 순서로 실행될지 예상할 수 있을까요? 🤔

제가 예상한 순서는 create_table > insert_data이지만 실제로는 테이블 형성이 안되어있는데 insert문이 먼저 호출되는 경우 실행이 불가하겠네요!!

찰리가 위에서 말씀주신 것처럼 insert문은 dao에서 직접 가능하도록 리팩터링 해주었습니다!

Gomding commented on src/main/java/chess/database/ConnectionGenerator.java:
```diff
@@ -0,0 +1,31 @@
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "root";
+    private static final Connection CACHE = makeConnection();
```
CACHE를 만드신 이유가 있을까요 🤔 

CACHE에 저장된 connection을 close했다면 어떤 일이 발생할까요?

coli-geonwoo replied:
cache가 반환되지만 사실 닫힌 커넥션이기에 ClosedConnectionException이 터질 것 같습니다!

사실 DB와 관해서 사용, 반환, 그리고 닫히면 어떻게 되는지에 대한 개념이 아직 모호하게 자리잡힌 것 같아요 😅 이번 리팩토링에서 다른 reference code를 참조해서 ConnectionPool을 통해 connection 생성 및 반환을 구현하면서 connection 개념을 이해하기 위해서 애썼습니다.

Gomding replied:
헉 Connection Pool 까지 구현해주셨군요 💯 💯 💯 
훌륭합니다..

Gomding commented on src/main/java/chess/database/ConnectionGenerator.java:
```diff
@@ -0,0 +1,31 @@
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
```
DB접속이 불가능한데 null을 return 하는것이 맞을까요..!

DB를 사용하는 서비스인데 DB 연결이 불가능하다면 치명적인 오류일것같아요 😱 
빠르게 에러를 던져서 바깥에 알리는것이 좋을것같아요

coli-geonwoo replied:
DB관련해서 수업시간에 사용한 code를 참고하여 코드를 작성하다보니 찰리에게 명확한 근거와 주관이 반영되어 있지 않은 코드를 전달드리게 되었네요. 죄송합니다. 🙇 

null이 반환되면 에러임에도 불구하고 에러 전달이 안될 수 있고 디버깅이 힘들어질 것 같아요.
책 <좋은 코드, 나쁜코드>에 있던 "에러는 최대한 빠르게, 그리고 요란하게 발생하라"라는 원칙과도 위배된다고 생각합니다. RuntimeException을 던져주도록 수정해주었습니다!

```java
public Connection createNewConnection() {
        try {
            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
        } catch (final SQLException e) {
            e.printStackTrace();
            throw new RuntimeException("DB 연결 오류");
        }
    }
```

Gomding replied:
에러 스택을 보고싶다면 `e.printStackTrace();` 를 사용하셔도 괜찮고
아래와 같이 RuntimeException의 두번째 인자를 활용해도 좋습니다 😄 

```java
        } catch (final SQLException e) {
            throw new RuntimeException("DB 연결 오류", e);
        }
```

coli-geonwoo replied:
RuntimeException 두번째 인자를 통해 printStackTrace가 가능한지 몰랐습니다! 꿀팁 감사합니다 찰리 👍

Gomding commented on src/main/java/chess/database/ConnectionGenerator.java:
```diff
@@ -0,0 +1,31 @@
+
+    public Connection getConnection() {
+        if (CACHE == null) {
+            return makeConnection();
```
CACHE는 makeConnection으로 인해 초기화 되는데

CACHE가 null이면 makeConnection() 도 null을 리턴하고 있지 않을까요..!

coli-geonwoo replied:
합리적인 지적이라 생각해요! 앞서 makeConnection에서 오류가 아닌 null을 리턴하면 발생하게 되는 문제가 바로 밑에 있었네요 🤔 

이번에는 CACHE가 아니라 connectionPool을 통해 커넥션 관리를 시도해보았습니다!! (이런 저런 refernece 참조해서 겨우 구현은 해내었지만 잘 작성된 코드인지는 아직 잘 모르겠습니다 😭 )

Gomding commented on src/main/java/chess/domain/board/Score.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.Objects;
+
+public final class Score {
```
Score 값 객체 사용 👍

Gomding commented on src/main/java/chess/domain/board/ScoreCalculator.java:
```diff
@@ -0,0 +1,68 @@
+
+public class ScoreCalculator {
+    private static final int MINUS_SCORE_THRESHOLD = 2;
+    private static final Score MINUS_SCORE = Score.from(0.5);
```
어떤것을 위한 MINUS_SCORE인지 이름에서 표현되면 좋겠어요!

coli-geonwoo replied:
PAWN_MINUS_SCORE로 rename 해주었습니다!

Gomding commented on src/main/java/chess/domain/board/ScoreCalculator.java:
```diff
@@ -0,0 +1,68 @@
+                .filter(piece -> piece.isSameTeam(team))
+                .map(Piece::getScore)
+                .reduce(Score::add)
+                .orElseThrow(() -> new IllegalStateException("팀의 기물이 하나도 없습니다."));
```
팀의 기물이 하나도 없다면 예외가 아니라 0점이 반환되는것이 의미적으로 맞지 않을까요..!

coli-geonwoo replied:
이 부분에 관해서는 조금 고민이 있었습니다.

제가 오류를 발생시킨 이유는
- 게임 승패는 체스게임이 끝났는지 여부에 따라 다르게 측정됩니다.
    - 진행중인 게임 : 점수를 계산해 더 높은 팀
    - 끝난 게임 : 킹이 살아있는 팀 
- 그렇기에 점수를 계산하는 상황에서 팀의 기물이 하나도 없는 상황은 체스 게임 내에서 문제가 있는 상황입니다.
- 따라서 점수를 계산해야 하는 runningGame임에도 기물이 없는 상황은 도메인 코드의 문제가 있는 상황이라 생각했고 오류를 던져주었습니다!

Gomding replied:
제가 잘못이해했을 수도 있지만..!
모든 기물이 다 잡힌 상태로 게임이 끝나는 경우도 있지않을까요? 🤔 (제일 마지막으로 King 이 잡히는 상황..?)
모든 기물이 다 잡힌 상태로 경기가 끝났을 때도 점수를 0점으로 계산해줘야하지 않을까~ 라는 생각이 들었습니다

```java
    @DisplayName("각 팀의 점수 합을 계산할 수 있다.")
    @Test
    void should_CalculateTeamScore() {
        Map<Position, Piece> board = Map.ofEntries(
                Map.entry(D3, new King(Team.BLACK)),
                Map.entry(F5, new Bishop(Team.BLACK))
        );
        ScoreCalculator calculator = ScoreCalculator.gameScoreCalculator();
        assertThat(calculator.calculateTeamScore(board, Team.WHITE)).isEqualTo(Score.from(0));
    }

```

coli-geonwoo replied:
헉, 예시까지 보여주셔서 감사합니다. 

우선 찰리의 리뷰를 반영해주었습니다! 
그러나, 제가 예외를 던진 이유에 대해 그 사고과정을 말씀드리면 다음과 같습니다.

EndGame일 경우 점수 계산이 호출되면 안된다고 생각했습니다.
경기 승패를 결정하는데 있어 점수가 낮음에도 이기는 경우들이 있기 때문입니다.

따라서, ChessGame 객체에서 경기 승패 판정로직을 endGame여부에 따라 나누었습니다.
RunningGame일 경우 => 점수 계산을 통해 높은 점수를 가진 사람이 승리
EndGame인 경우 => king이 살아있는 팀이 승리

```java
public Team winTeam() {
        if (isEndGame()) {
            return winTeamInEndGame();
        }
        return winTeamInRunningGame();
    }

    private Team winTeamInEndGame() {
        if (board.isKingAlive(Team.BLACK)) {
            return Team.BLACK;
        }
        return Team.WHITE;
    }

    private Team winTeamInRunningGame() {
        Score whiteScore = board.teamScore(Team.WHITE);
        Score blackScore = board.teamScore(Team.BLACK);
        if (whiteScore.isHigherThan(blackScore)) {
            return Team.WHITE;
        }
        if (blackScore.isHigherThan(whiteScore)) {
            return Team.BLACK;
        }
        return Team.NONE;
    }
```

만약 king이 잡혔음에도 (기물이 하나도 없음에도) 점수 계산을 호출하여 승패를 판단하는 행위는 도메인 규칙에 어긋나는 승패판정이라 생각하여 오류를 던져주었습니다.

그럼에도 제가 **찰리의 리뷰를 반영한 이유**는 이펙티브 자바 `아이템69. 예외는 진짜 예외 상황에만 던져주어라`라는 책의 내용으로부터 찰리의 제안이 합리적이라 생각했기 때문입니다.  예외를 단순 흐름 제어용으로 사용하고, 이를 호출하는 코드 사용자에게도 강요하는 것은 좋지 않다는 것에서 기물이 없으면 예외를 반환하는 행위가 좋지 않음을 인식할 수 있었습니다!

Gomding replied:
아하 이 부분은 제가 코드 이해가 부족했네요 🥲
승리 시(King 이 잡힌 경우)는 점수 계산을 하지 않고 있었군요

그렇다면..! 콜리의 예외를 던지는것이 이해가 됩니다 😢 
콜리의 의견대로 예외를 던지는것이 더 맞다고 생각해요!

Gomding commented on src/main/java/chess/domain/piece/Rook.java:
```diff
@@ -10,9 +11,10 @@
 
 public class Rook extends Piece {
     private static final List<Direction> PASSING = List.of(UP, DOWN, RIGHT, LEFT);
+    private static final Score score = Score.from(5);
```
각 기물이 점수를 가지게 하신것 👍

Gomding commented on src/test/java/chess/dao/BoardDaoTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Test
+    void should_FindBoard() {
+        ChessBoard board = ChessBoard.normalBoard(Map.ofEntries(Map.entry(D3, Pawn.of(Team.BLACK))));
+        FakeBoardDao boardDao = FakeBoardDao.runningBoard(board.status());
```
BoardDao 를 테스트한다는것은 어떤것을 의미할까요?
데이터 접근 객체이니 실제 데이터 접근을 테스트해봐야될 것 같아요!

테스트를 위해 만든 FakeBoardDao를 검증하는것에 어떤 의미가 있을까요? 🤔 

테스트를 위한 가짜 객체를 사용하는 것이 무엇을 위한 것일지 고민해보면 좋겠어요!
우선 ChessGameService 에 대한 테스트를 작성해보시죠..!
그리고 Service 객체 테스트에 실제 DB가 필요한지 Stub으로 충분할지 고민해볼 수 있겠어요~

Gomding commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,49 @@
+import chess.domain.game.ChessGame;
+import chess.domain.position.Position;
+
+public class ChessGameService {
```
오 Service 계층과 객체를 만들어 주셨군요
어떤 필요로 만들어주셨을까요?

coli-geonwoo replied:
찰리의 리뷰를 바탕으로 이런저런 자료를 많이 찾아보았는데요! (사실 Layer 계층 구조에 대한 배경지식이 없어 어떤 계층이 어떤 역할을 하는지에 대해 얄팍한 이해를 하고 있는 상황입니다 😭 )

그래서 차근차근 정리해보니 다음 과 같은 해결사안들이 있었습니다.
- controller에서 dao를 직접 알지 못하게 캡슐화할 필요성
- dao와 도메인 간의 상호작용을 담당할 인터페이스 계층의 필요성

그리고, 이 두 역할에 알맞은 계층이 Service 계층임을 알게되었고 ChessGameService를 만들어주었습니다. 이로 부터 얻은 효용은 다음과 같았습니다.

**-Service객체 안으로 각각의 트랜잭션 로직 캡슐화 >> controller의 run()메소드 단순화 됨**
   
**- 코드 사용자 입장에서 dao와의 소통을 더욱 쉽게 판별 가능해짐**

Gomding replied:
아마 Service 계층은 다음 레벨에서 충분히 고민해보실 수 있을거에요 :)
Domain 이 비즈니스 로직을 담당한다면 
Service 는 비즈니스의 흐름제어를 담당한다고 해요~

Gomding commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,49 @@
+import chess.domain.position.Position;
+
+public class ChessGameService {
+    private final DBServcie dbServcie;
```
DBService 가 꼭 필요까요..!
ChessGameService 내부에서 BoardDao, TurnDao 를 가져도 될것같아요 😃 

ChessGameService와 DBService 는 각각 어떤 책임을 가지고있을까요?

coli-geonwoo replied:
사실, `인스턴스변수를 2개 까지로 한다`라는 요구사안을 최대한 만족시켜주고 싶어서 boardDao와 turnDao간의 상호작용을 담당해주는 DBService를 만들었습니다.

- ChessGameService : playTurn에 따른 보드 정보 업데이트, dao와의 상호작용 캡슐화를 통한 controller로부터 각 트랜잭션 정보 캡슐화
- DbService : boardDao와 turnDao을 이용해 ChessGame 저장 및 찾기 기능 제공

다만 `Dao와 Domain간의 트랜잭션을 관리한다`라는 의미의 Service 계층이라면 DbService와 ChessGameService간의 역할 중복이 있는 것 같습니다. 

기존의 boardDao와 turnDao을 활용해 ChessGame 단위의 저장 및 찾기 기능을 제공했던 책임을 ChessGameService가 책임지도록 리팩터링해주었습니다!

Gomding replied:
훌륭하네요! 👍 

필요하다면 나눠줘도 괜찮지만 인스턴스변수를 2개로 해야한다는 강제에 의해서라면 
굳이 분리하지 않아도 된다고 생각해요 :)

인스턴스 변수 2개로 제한하는 것은 너무 많은 책임을 가지지 않게 하려는 의도이고
현재는 ChessGameService 의 책임이 과한 상황은 아니라고 생각했어요 😄

Gomding commented on src/main/java/chess/dao/DBServcie.java:
```diff
@@ -0,0 +1,40 @@
+import chess.domain.piece.Piece;
+import chess.domain.position.Position;
+
+public final class DBServcie {
```
이런 경우 카멜케이스(파스칼 케이스?)라면 `DbService` 로 이름을 지어주시는게 좋아요!

약어의 경우 대문자로 전부 써버리면 오히려 혼란을 줄 수 있어요 😃 
예를들면 `HTTPWWW` 의 경우 약어를 모르는 사람이면 어떤 단어가 조합된건지 모를 수 있어요
`HttpWww` 라면 명확하게 구분이 가능하죠!

coli-geonwoo replied:
찰리의 예시를 보고 단번에 이해했습니다! 친절한 설명까지 감사합니다 😄 
기존 ConnectionGenerator를 이번에 DbConnector로 rename 해주었는데 카멜케이스를 신경써보았습니다 :)

Gomding reviewed (changes requested):
안녕하세요 콜리!
아직 서비스 실행하기가 쉽지 않네요 😢 
관련해서 코멘트 남겼으니 확인 부탁드려요 🙏 

궁금한 점 있으면 DM 으로 직접 물어봐주세요!

coli-geonwoo committed:
docs: 3-4단계 2차 리뷰 TODO 작성

coli-geonwoo committed:
rename: minus 점수의 의도가 드러나도록 구체화

MINUS_SCORE_THRESHOLD > PAWN_MINUS_SCORE_THRESHOLD
MINUS_SCORE > PAWN_MINUS_SCORE

coli-geonwoo committed:
refactor: DbService의 책임을 ChessGameService로 이전

coli-geonwoo committed:
refactor: DB연결 실패 시 RuntimeException throw

coli-geonwoo committed:
refactor: 테이블 초기화를 dao에서 처리

coli-geonwoo committed:
refactor: 테이블 초기화를 dao에서 처리

coli-geonwoo committed:
rename: ConnectionGenerator > DbConnector

coli-geonwoo committed:
refactor: 데이터가 있는지 확인하는 코드 단순화

coli-geonwoo committed:
feat: 커넥션 풀 구현

coli-geonwoo committed:
test: ChessGameService 작성

coli-geonwoo committed:
refactor: 초기 insert문 sql 삭제

coli-geonwoo committed:
refactor: FakeBoardDao, FakeTurnDao를 통한 service 계층 test구현

coli-geonwoo committed:
docs: 2차 개선 사안 작성

coli-geonwoo commented:
안녕하세요 찰리 🙌 

3-4단계 2차 개선사안에 대해 말씀드립니다.

---
### 2차 리뷰 개선사안

#### domain
- rename : MINUS_SCORE > PAWN_MINUS_SCORE
- DbService 삭제 : ChessGameService로 책임 이전
- database 
  - Dbconnector : connection 연결 책임
  - ConnectionPool : connectionPool 구현
- db
  - BoardDaoImpl : 초기 테이블 insert문 구현
  - TurnDaoImpl : 초기 턴 insert문 구현

### test
- FakeBoardDao와 FakeTurnDao를 활용한 ChessGameServiceTest 구현
---

유의한 변화는 **ConnectionPool 구현과 Service계층 테스트를 위한 FakeBoardDao , FakeTurnDao의 활용**입니다.
커넥션 풀의 경우 배경지식이 없어 reference를 참조하며 꾸역꾸역 구현해보았는데요. 그 과정에서 이해되는 지점들이 많았고 커넥션 자원의 부하가 상당하다는 점, 그리고 오랜시간 호출되지 않으면 연결이 자동해제된다는 점 등의 커넥션 특징을 익힐 수 있었습니다.

이전의 드렸던 테스트 더블과 관련된 의문들은 직접 구현하면서 어느정도 해소된 것 같습니다.
다만 제가 내린 결론들이 알맞은 결론들인지 찰리에게 확인받고 싶은 마음입니다.

**> 배운점1. 우리가 테스트해야하는 것은 도메인 영역이다.**
- 단순 데이터 접근 로직 / view 로직이 아니라, 도메인과 연관있는 service계층의 테스트가 유의하다

---

**>배운점2. 테스트 더블의 의의는 실제 DB에 영향을 주지 않는 것이다.**
- FakeObject를 통해 같은 로직을 수행하되 Db에 영향을 주지 않음으로서 안전하게 테스트를 수행가능하다

---

**> 배운점3. 커넥션의 경우, 오랜시간 호출되지 않으면 자동해제 된다.**
- 커넥션이 사용가능한지 실험 쿼리를 실행해보고 만약 연결이 끊어졌다면 다시 재연결해 반환해야 한다.

---

**> 배운점4. 오류의 경우 호출 대상의 복구 가능성을 고려해야 한다.**
- SQLException의 경우 애플리케이션 레벨에서 복구가 불가하다
- 무의미한 throws를 막기 위해서 UncheckedException으로의 포장을 해주면 좋다.


---

**번외로 미션이 마무리되는 시점에서 찰리께 드리고픈 감사의 말씀**

매번 애정을 가지고 코드 살펴봐 주셔서 감사드립니다. 감사하다는 말을 자주 반복하게 되는데, 사실 제 마음을 표현드릴 수 있는 말이 이것밖에 없네요 😅   다른 미션들보다 조금 버거운 미션이었는데 이번 chess미션에서 찰리 리뷰 하나하나를 복기하며, 리플렉션, instanceOf의 안좋은점, 개발자가 바라보아야 하는 예외처리의 관점, 추상화와 다형성의 활용성, DB 연결 기초 등 많은 걸 얻어갑니다!! 찰리가 제게 좋은 리뷰어였던 것처럼 저도 찰리에게 좋은 리뷰이였었다면 좋겠어요 😄 

부족한 점이 있다면 request Change주시면 적극 반영해보겠습니다 🙇

Gomding commented on src/main/java/chess/dao/BoardDaoImpl.java:
```diff
@@ -0,0 +1,141 @@
+    public BoardDaoImpl(ConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+        if (isFirstCall()) {
+            initializeBoard();
```
왜인지 모르겠지만 board 초기화 안되더라구요 😢

coli-geonwoo replied:
isFirstCall에 대한 쿼리문이 잘못되어 있었습니다.. 중간 로직을 수정하고 application을 돌리지 않아보았네요 😓 초기화 오류 해결해보았습니다! 이번엔 꼭 어플리케이션이 DB와 잘 연계되어 돌아갔으면 좋겠네요 😭

Gomding replied:
잘 해결해주셨군요 😃 
지금은 잘 실행됩니다 🎉

Gomding commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,31 @@
+    team varchar(20) null,
+    constraint turn
+        foreign key (team) references team_color (color)
+);
```
사실 team_color, piece_type 같은 경우는 애플리케이션 실행을 위해 꼭 필요한 값들이니

여기 sql 파일에서 insert 해주시는게 좋습니다!

해당 값이 없어서 애플리케이션 실행하는데 조금 힘들었어요 😅

coli-geonwoo replied:
외래키의 부작용들이 많다는 찰리의 리뷰를 듣고 스키마를 더 간단히 바꾸었습니다! 이번 PR에는 수정된 스키마로 보내드리겠습니다!

```java
create table board
(
    position       varchar(20) not null
    primary key,
    distinct_piece binary(1)   not null,
    piece_type     varchar(20) null,
    team           varchar(20) null
);

create table turn
(
    team varchar(20) null
);
```

Gomding commented on src/main/java/chess/dao/BoardDaoImpl.java:
```diff
@@ -0,0 +1,141 @@
+    public BoardDaoImpl(ConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+        if (isFirstCall()) {
+            initializeBoard();
```
제가 의도했던것은 체스 로직에서 자연스럽게 새로운 Board 가 저장되는 것이었어요~

예를들면 ChessGameService 에서 ChessGame 생성 시 
- 현재 저장된 게임이 없다면 보드를 새로 생성하여 저장
- 기존에 보드가 있고 게임중이면 조회
- 기존 게임이 끝났다면 새로운 게임 시작

++) 현재처럼 하나의 Board만 운영된다면 로컬의 싱글 게임으로 끝날것 같은데요
방학동안 심심하다면 여러개의 체스게임방을 운영할 수 있도록 리팩터링 해보는것도 좋을것같아요~
그러기 위해서는 Game 방에대한 데이터도 있어야할것같구요~ 😄

coli-geonwoo replied:
이 부분에서 질문이 있습니다!

현재 게임애플리케이션의 로직은 다음과 같습니다. (싱글 게임)

- db의 보드가 안끝났다면 => 기존의 보드로 service 객체 생성
- 기존의 보드가 끝났다면 => 새로운 게임보드로 service 객체 생성
- 기존 게임이 끝났다면 => 애플리케이션 종료

game 객체에서 board는 되도록 final로 유지해주고 싶어 board를 바꾸지 않고, 애플리케이션 실행 하나당 하나의 보드로 진행하고자 했습니다. 

저는 관련 로직을 처리해주었다고 생각하는데 찰리가 전달하고자 하는 의도가 어떤 부분인지 혹시 더 자세한 설명을 부탁드려도 될까요?? 😶  

---
> 방학동안 심심하다면 여러개의 체스게임방을 운영할 수 있도록 리팩터링 해보는것도 좋을것같아요~

이 부분도 방학기간 때 더 다루어볼 생각입니다. 지금은 gameID나 db schema가 확장에 좋은 구조는 아니라고 생각해서 우선 싱글게임 애플리케이션을 잘 돌아가게 구현하고,  스스로 애플리케이션을 더 확장해보겠습니다!

Gomding replied:
ChessGameController를 보면 아직 start 명령어를 입력하지 않았음에도 Board데이터가 저장되고 있어요
`누군가 Board 가 첫번째로 생성되는 부분은 어디지?` 라고 생각하고 추측하면 어디를 볼까요?
Controller 로직? ChessGameService 객체 생성 로직? 
저는 게임이 start 되는 시점을 제일 먼저 살펴봤을것같아요. (실제로도 그렇게 했구요!)
OOP(객체 지행 프로그래밍)를 하면 예측 가능한 코드를 작성한다는 것이 장점인데
그 말은 꼭 OOP가 아니더라도 `예측 가능한 코드를 작성한다는 것` 이 중요하다는 말이 되겠죠?

(물론 사람마다 예측하는 방향은 다를거라 생각해요 😃 )

```java
// ChessGameController.java
    public void run() {
        Command command;
        ChessGameService gameService = ChessGameService.of(BoardDaoImpl.of(), TurnDaoImpl.of());
        ChessGame game = startChessGame(gameService);

        do {
            command = retryUntilNoError(this::readCommand);
            command.execute(gameService, outputView);
        } while (!command.isEnd() && !gameService.isEndGame());

        outputView.printStatusMessage(game);
    }
```

Gomding commented on src/main/java/chess/database/ConnectionPool.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Set;
+import java.util.Stack;
+
+public class ConnectionPool {
```
Connection Pool 구현 👍

Gomding commented on src/main/java/chess/database/ConnectionPool.java:
```diff
@@ -0,0 +1,90 @@
+
+    public Connection getConnection() throws SQLException {
+        if (isFull()) {
+            throw new RuntimeException("커넥션 풀이 모두 차있습니다");
```
**절대 반영하지 않아도 되는 코멘트**
커넥션 풀의 모든 커넥션을 사용중이라면 대기시간을 가져도 좋을것같아요~
최대 10초정도 1초마다 커넥션 가져오는걸 시도하고 그래도 없으면 예외를 던져주면 좋을것같습니다 😄

coli-geonwoo replied:
한번 시도해보았습니다!! 그러나, db와 관한 이해도가 낮아 찰리가 대기시간을 가지는 것을 제안주신 이유가 궁금합니다 🤔 저는 connection 반납에도 시간이 걸리기 때문에, 충분한 시간을 두고 connection 연결을 시도하는 것이 바람직하다는  의도로 이해했습니다.

```java
  public synchronized Connection getConnection(int tryCnt) throws SQLException {
        if (isAllConnectionUsed()) {
            checkRetryCount(tryCnt);
            waitOneSeconds();
            return getConnection(tryCnt + 1);
        }

        Connection connection = getConnectionFromPool();
        if (connection == null) {
            connection = createNewConnectionForPool();
        }

        return makeAvailable(connection); //일정 시간이 지나 끊긴 connection을 다시 연결 후 반환
    }

 private void checkRetryCount(int tryCnt) {
        if (tryCnt == MAX_RETRY_COUNT) {
            throw new RuntimeException("커넥션 풀이 모두 차있습니다");
        }
    }

private void waitOneSeconds() {
        try {
            TimeUnit.SECONDS.sleep(1);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

```

Gomding replied:
모든 커넥션이 사용중이라 커넥션을 못받는다는 이유로
특정 요청은 분명 정상적인데도 커넥션을 얻지못해 에러가 발생하는 아이러니한 상황이 발생할 거에요 😢 

하지만 작게는 밀리초 많게는 몇초 정도 기다리면 그래도 커넥션을 얻어서 정상적인 로직 실행이 가능하죠
(대부분의 로직은 커넥션을 얻어서 1~3 초안에는 커넥션을 반납할거에요 ㅎㅎ)

그래서 일정시간을 기다리면서 커넥션을 요청하면 에러도 훨씬 줄어들겠죠?
사용자 경험도 좋아질 것이구요!

물론 지금은 싱글 게임이라 크게 의미있는 로직은 아닙니다 ㅎㅎ

Gomding commented on src/main/java/chess/database/ConnectionPool.java:
```diff
@@ -0,0 +1,90 @@
+        return connection;
+    }
+
+    private boolean isFull() {
```
isFull 보다는 `커넥션을 모두 사용중임` 이라는 의미가 들어가면 좋곘어요!

coli-geonwoo replied:
isAllConnectionUsed로 rename 해주었습니다!

Gomding replied:
훌륭한 이름이네요 👍

Gomding commented on src/test/java/chess/service/FakeBoardDao.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class FakeBoardDao implements BoardDao {
```
해당 객체는 어디 패키지가 어울릴까요! 😄 
service 말고 더욱 적절한 패키지가 있을것같아서요..!

coli-geonwoo replied:
chess/service > chess/domain/dao/fakedao로 패키지를 이동해 주었습니다.

Gomding commented on src/test/java/chess/service/ChessGameServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+    @DisplayName("기물 이동을 통해 턴을 진행할 수 있다")
+    @Test
+    void should_PlayTurn_When_GiveStartAndDestination() {
+        OutputView outputView = new OutputView();
```
OutputView가 Sevice 테스트 시 꼭 필요한 객체일까요?

coli-geonwoo replied:
전체적인 기물 상황을 판단하는데 있어 outputView를 통해 판단할 수 있다 생각했는데, 밑에 찰리가 주신 리뷰처럼 게임의 상태판단을 통해서도 테스트가 가능할 것 같습니다. 또 그 방향이 더 합리적이라 생각하기도 하고요!

outputView가 service test 내에서 활용되지 않도록 리팩터링해주었습니다.

Gomding commented on src/test/java/chess/service/ChessGameServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+        String expectedGame = outputView.printChessBoardMessage(movedGame.getBoard());
+        String actualGame = outputView.printChessBoardMessage(service1.gameBoard());
+
+        assertThat(actualGame).isEqualTo(expectedGame);
```
아래 처럼 검증을 해보면 어떨까요!
ChessGameService 에서 검증하고 싶은 상태를 검증해보는 것이죠 :)

```java
    @DisplayName("기물 이동을 통해 턴을 진행할 수 있다")
    @Test
    void should_PlayTurn_When_GiveStartAndDestination() {
        ChessGame newGame = ChessGame.newGame();
        FakeBoardDao boardDao = new FakeBoardDao(newGame.getBoard());
        FakeTurnDao turnDao = new FakeTurnDao(newGame.getTurn());
        ChessGameService service = new ChessGameService(boardDao, turnDao, newGame);
        ChessBoard board = boardDao.findBoard();
        Piece piece = board.status().get(D2);


        service.playTurn(D2, D4);

        ChessBoard movedBoard = boardDao.findBoard();
        Piece movedPiece = movedBoard.status().get(D4);
        assertThat(piece).isEqualTo(movedPiece);
    }
```

Gomding commented on src/test/java/chess/service/FakeBoardDao.java:
```diff
@@ -0,0 +1,47 @@
+    public void updatePiecePosition(Position position, Piece piece) {
+        Map<Position, Piece> piecePositons = board.status();
+        piecePositons.put(position, piece);
+        this.board =  this.board = new ChessBoard(piecePositons, ScoreCalculator.gameScoreCalculator());
```
여긴 오타가 있군요 :)

coli-geonwoo replied:
수정해주었습니다!

Gomding commented on src/test/java/chess/service/ChessGameServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+    void should_SaveNewGame_When_IsEndGame(){
+        OutputView outputView = new OutputView();
+        ChessGame newGame = ChessGame.newGame();
+        ChessGameService service = new ChessGameService(new FakeBoardDao(newGame.getBoard()), new FakeTurnDao(newGame.getTurn()), newGame);
```
ChessGame, ChessGameService 초기화 하는 부분이 반복되니 Setup 함수에서 초기화해봐도 되겠어요~

coli-geonwoo replied:
@BeforeEach를 통해 중복코드를 제거해주었습니다!!

Gomding reviewed (changes requested):
안녕하세요 콜리!
고민의 흔적이 많이 보이네요 👍 

몇가지 코멘트 남겼으니 확인 부탁드려요~

coli-geonwoo committed:
fix: 보드 초기화 오류 해결

coli-geonwoo committed:
style: 변수명 오타 수정

coli-geonwoo committed:
docs: 3차 개선사안 TODO 작성

coli-geonwoo committed:
rename: 모든 커넥션이 사용중이라는 의미가 잘 드러나도록 메서드명 변경

isFull > isAllConnetionUsed

coli-geonwoo committed:
refactor: 팀의 기물이 없는 경우 0점을 반환하도록 개선

coli-geonwoo committed:
test: setUp함수를 통한 중복 코드 분리

coli-geonwoo committed:
chore: test용 fakeDao 패키지 이동

test/service > test/dao/fakedao

coli-geonwoo committed:
chore: 완료한 TODO 삭제

coli-geonwoo committed:
refactor: outputview 메서드 반환형 변환

String > void

coli-geonwoo committed:
refactor: connectionPool이 모두 차있을 때 기다리는 기능 추가

coli-geonwoo committed:
refactor: 스트림식 단순하게 수정

coli-geonwoo committed:
docs: 3-4단계 3차 개선사안 작성

coli-geonwoo commented:
안녕하세요 찰리 🙌 

이제 정말 미션 막바지에 가까워가네요!! 그런데 저는 왜 이렇게 아쉬울까요 😂  많이 부족한 만큼 열심히 노력했지만 잠시 시간을 가졌다 코드를 바라보면 여기저기 부족한 점이 계속 보이는 것 같습니다.

db 오류가 계속 나 이번에는 꼭! 찰리 로컬에서 잘 실행이 되었으면 좋겠는데요
수정한 db schema와 오류 원인이었던 쿼리문을 수정하여 다시 PR드립니다. 제 로컬에서는 잘 실행되는데 찰리 로컬에서도 이번에는 꼭 실행되었으면 좋겠어요 😭 

3-4단계 3차 개선사안 전달드립니다!!

### 3차 리뷰 개선사안

#### domain

- dao : 보드 및 턴 초기화 오류 해결
- ScoreCalculator: 팀의 기물이 없는 경우 0점을 반환하도록 변경
- database/ConnectionPool : Connection이 모두 사용중일 때 기다리는 기능 추가

#### test

- test용 fakeDao 패키지 이동 : test/service > test/dao/fakedao
- setUp 함수를 통한 중복 코드 분리
- outputView가 test 내에 사용되지 않도록 의존성 제거

---

감사합니다 🙇

Gomding commented on src/main/java/chess/controller/command/Command.java:
```diff
@@ -1,10 +1,11 @@
+        permits StartCommand, MoveCommand, StatusCommand, EndCommand {
+    void execute(ChessGameService service, OutputView outputView);
 
     boolean isStart();
```
Command 객체가 InputView, OutputView 도 필드로 가진다면
isStart 같은 메서드가 없어도 될것같아요 :)

execute 메서드 하나도 퉁칠 수 있지 않을까요~

Gomding commented on src/main/java/chess/database/DbConnector.java:
```diff
@@ -0,0 +1,22 @@
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("DB 연결 오류");
```
`RuntimeException` 생성자의 두 번째 인자(cause)를 활용하면 `e.printStackTrace` 를 호출하지 않아도 될것같아요~ 😄 
예외 출력 시 RuntimeException의 에러 스택에 cause가 함께 나옵니다 😉

`throw new RuntimeException("DB 연결 오류", e);`

![image](https://github.com/woowacourse/java-chess/assets/57378410/9b78a772-a4bb-48ba-9400-a2cf122a9c64)

(위 에러는 제가 강제로 발생시켰어요~)

Gomding commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,70 @@
+
+    public void end() {
+        if (game.isEndGame()) {
+            saveGame(ChessGame.newGame());
```
여기도 `게임이 끝났는데 게임을 새로만든다?` 라는것이 혼란을 줄 수 있어보여요 :)

Gomding reviewed (approved):
안녕하세요 콜리!
이제 남길 코멘트가 없네요 👏 👏 
체스 미션 고생많으셨어요 :)

방학 즐겁게 보내시고 레벨2도 화이팅입니다~!! 🔥 🔥
